spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy

```diff
@@ -0,0 +1,1241 @@
+    if (!outEndsWith(NEW_LINE)) {
+      trimSpaceRight()
+      _out.print '\n'
+    }
+    readyToIndent = true
+  }
+
+  def printDoubleBreak() {
+    if (outEndsWith('\n\n')) {
+      // do nothing
+    } else if (outEndsWith(NEW_LINE)) {
+      _out.print '\n'
+    } else {
+      trimSpaceRight()
+      _out.print '\n'
+      _out.print '\n'
+    }
+    readyToIndent = true
+  }
+
+  void visitPackage(PackageNode packageNode) {
+
+    if (packageNode) {
+
+      showOn(Show.ANNOTATIONS) {
+        packageNode.annotations?.each {
+          visitAnnotationNode(it)
+          printLineBreak()
+        }
+      }
+
+      if (packageNode.text.endsWith('.')) {
+        print packageNode.text[0..-2]
+      } else {
+        print packageNode.text
+      }
+      printDoubleBreak()
+    }
+  }
+
+  void visitImport(ImportNode node) {
+    if (node) {
+      showOn(Show.ANNOTATIONS) {
+        node.annotations?.each {
+          visitAnnotationNode(it)
+          printLineBreak()
+        }
+      }
+      print node.text
+      printLineBreak()
+    }
+  }
+
+  @Override
+  void visitClass(ClassNode node) {
+
+    classNameStack.push(node.name)
+
+    showOnAll(EnumSet.of(Show.ANNOTATIONS, Show.CLASS)) {
+      node?.annotations?.each {
+        visitAnnotationNode(it)
+        printLineBreak()
+      }
+    }
+
+
+    showOn(Show.CLASS) {
+      visitModifiers(node.modifiers)
+      if (node.interface) print node.name
+      else print "class $node.name"
+      visitGenerics node?.genericsTypes
+      print ' extends '
+      visitType node.unresolvedSuperClass
+      boolean first = true
+      node.unresolvedInterfaces?.each {
+        if (first) {
+          print ' implements '
+        } else {
+          print ', '
+        }
+        first = false
+        visitType it
+      }
+      print ' {'
+      printDoubleBreak()
+    }
+
+    readyToIndent = true
+
+    indented {
+      showOn(Show.PROPERTIES) {
+        node?.properties?.each { visitProperty(it) }
+        printLineBreak()
+      }
+      showOn(Show.FIELDS) {
+        node?.fields?.each { visitField(it) }
+        printDoubleBreak()
+      }
+      showOn(Show.CONSTRUCTORS) {
+        node?.declaredConstructors?.each { visitConstructor(it) }
+        printLineBreak()
+      }
+      showOn(Show.OBJECT_INITIALIZERS) {
+        visitObjectInitializerBlocks(node)
+        printLineBreak()
+      }
+      showOn(Show.METHODS) {
+        node?.methods?.each { visitMethod(it) }
+      }
+    }
+    showOn(Show.CLASS) {
+      print '}'
+      printLineBreak()
+      classNameStack.pop()
+    }
+  }
+
+  private void visitObjectInitializerBlocks(ClassNode node) {
+    for (Statement stmt : (node.objectInitializerStatements)) {
+      print '{'
+      printLineBreak()
+      indented {
+        stmt.visit(this)
+      }
+      printLineBreak()
+      print '}'
+      printDoubleBreak()
+    }
+  }
+
+  private void visitGenerics(GenericsType[] generics) {
+
+    if (generics) {
```

Comment:
```suggestion
    if (generics != null) {
```

Reply:
why the explicit `!= null` instead of simple groovy truth?

Reply:
Because it is a different case.
`ArrayList<>` for example.
With `if (generics)` it is `ArrayList` as the `generics` array is empty and thus Groovy falsy.
Wiht `if (generics != null)` it is rendered correctly as `ArrayList<>`.